src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -441,36 +449,34 @@ private static bool ContractMatches(DataContract contract, DataContract cachedCo
 
             internal static int GetId(RuntimeTypeHandle typeHandle)
             {
-                lock (s_cacheLock)
+                var typeHandleRef = GetThreadLocalTypeHandleRef(typeHandle);
+
+                try
                 {
-                    IntRef? id;
-                    typeHandle = GetDataContractAdapterTypeHandle(typeHandle);
-                    s_typeHandleRef.Value = typeHandle;
-                    if (!s_typeToIDCache.TryGetValue(s_typeHandleRef, out id))
+                    var id = s_typeToIDCache.GetOrAdd(typeHandleRef, static (key) =>
                     {
-                        int value = s_dataContractID++;
-                        if (value >= s_dataContractCache.Length)
+                        lock (s_cacheLock)
```

Comment:
Just an observation: under load, you'll create unused entries in **s_dataContractCache** as **ConcurrentDictionary<TKey,TValue>** does not guarantee that the specified delegate is only invoked once for a given key.

Reply:
That is a good observation.
It think something should be done about that commit.

I wonder if it makes sense to use Lazy to protect against that or do something similar to "double checked locking" (using TryGet before lock) or just use a reader/writer lock and fall back to dictionary. 

I am open to drop that commit for know and maybe open a separate pr at a later time after comparing the alternative fixes

Reply:
I did a small benchmarks (single threaded), which just invokes GetId for different keys with almost no adds just to measure overhead for the non-contented case and settled for using concurrent dictionary + lazy.

``` ini

BenchmarkDotNet=v0.13.1, OS=Windows 10.0.22000
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK=7.0.100-preview.4.22252.9
  [Host]     : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT
  DefaultJob : .NET 6.0.5 (6.0.522.21309), X64 RyuJIT


```
|     Method |     Mean |     Error |    StdDev |
|----------- |---------:|----------:|----------:|
|   Original | 2.380 μs | 0.0126 μs | 0.0118 μs |
|       Lazy + Concurrent (third commit)| 3.076 μs | 0.0208 μs | 0.0195 μs |
|     Faulty (second commit)  | 3.111 μs | 0.0369 μs | 0.0327 μs |
| ReaderWriterLockSlim and dictionary| 5.021 μs | 0.0192 μs | 0.0180 μs |